include/rsp.inc

```diff
@@ -0,0 +1,680 @@
+                vmudm \vdstreg, \vsrcreg, \vshiftreg,11
+            .elseif (\qty == 5)
+                vmudm \vdstreg, \vsrcreg, \vshiftreg,12
+            .elseif (\qty == 6)
+                vmudm \vdstreg, \vsrcreg, \vshiftreg,13
+            .elseif (\qty == 7)
+                vmudm \vdstreg, \vsrcreg, \vshiftreg,14
+            .elseif (\qty == 8)
+                vmudm \vdstreg, \vsrcreg, \vshiftreg,15
+            .elseif (\qty >= 9 && \qty <= 15)
+                .error "Use vsra8 for quantities in range 9-15"
+            .else
+                .error "Invalid quantity in vsra"
+            .endif
+        .endm
+    .endm
+.endm
+
+##################################################
+# Other psuedo-ops
+##################################################
+
+# jal_and_j implements a tail-call-optimized subroutine call followed by
+# a jump. In other words, this:
+#
+#   jal_and_j func, label
+#
+# is equivalent to:
+#
+#   jal func
+#   nop
+#   j label
+#   nop
+#
+# but using only two instructions instead of four.
+.macro jal_and_j func, label
+    j \func
+    li ra, %lo(\label)
+.endm
+
+
+##################################################
+# Hardware registers
+##################################################
+
+#define SP_STATUS_BROKE         (1<<0)
+#define SP_STATUS_DMA_BUSY      (1<<2)
+#define SP_STATUS_DMA_FULL      (1<<3)
+#define SP_STATUS_SIG0          (1<<7)
+#define SP_STATUS_SIG1          (1<<8)
+
+#define SP_WSTATUS_RESET_SIG0   (1<<9)
+#define SP_WSTATUS_SET_SIG0     (1<<10)
+#define SP_WSTATUS_RESET_SIG1   (1<<11)
+#define SP_WSTATUS_SET_SIG1     (1<<12)
+
+#define COP0_DMA_SPADDR         $0
+#define COP0_DMA_RAMADDR        $1
+#define COP0_DMA_READ           $2
+#define COP0_DMA_WRITE          $3
+#define COP0_SP_STATUS          $4
+#define COP0_DMA_FULL           $5
+#define COP0_DMA_BUSY           $6
+
+#define COP2_CTRL_VCO           $0
+#define COP2_CTRL_VCC           $1
+#define COP2_CTRL_VCE           $2
+
+
+##################################################
+# RSP_DMA.INC macros
+##################################################
+
+# Compose the RSP DMA size register with immediate constants "width" and "height".
+# For a linear transfer, use height=1.
+# Notice that if height!=1, a pitch must be specified in the register. This
+# version of the macro is useful for the common case in which the pitch is not
+# a compile-time constant and thus will need to be inserted at runtime.
+#define DMA_SIZE(width, height)          (((width)-1) | (((height)-1)<<12))
+
+# Compose the RSP DMA size register with immediate constants
+# "width", "height", and "pitch".
+# It is similar to DMA_SIZE, but it can be used when the pitch is a compile-time
+# constant.
+#define DMA_SIZE3(width, height, pitch)  (((width)-1) | (((height)-1)<<12) | (((pitch)-(width))<<20))
+
+# DMAExec operation macros. Pass one of these in t2 to DMAExec
+#define DMA_IN_ASYNC     0x00000000
+#define DMA_OUT_ASYNC    0xFFFF8000
```

Comment:
Was the `8000` part an arbitrary choice?

Reply:
Semi-arbitrary. We need a 16-bit negative value that sign-extends to a 32-bit negative value, so that you can load it with non-macro `li` (which does a 16-to-32 sign extension). This basically allows to load all the `DMA_` operation macros into `t2` with a single `li` opcode.

`0x8000` would work just as `0xCCCC`, but it leaves 15 free bits to encode other things. In fact, we do encode the wait flags in the lowest bits (we reuse the same bits of the status register to simplify the checking loop).

Reply:
I have seen the part where we used the lowest bits and its ingenious but I thought `0xFFFF0000` would also work?

Reply:
Ah sorry we use 16bits